### PR TITLE
chore(samples): Snippet Generating System (part 1)

### PR DESCRIPTION
**Pull request by m-strzelczyk**
#### Why did it happen?
After working on code samples for the Python GAPIC library for couple months now, I have learned two things:

1. Most of the snippets will be really similar to each other. After all, most of them are about creating a Virtual Machine one way or another with just some minor adjustments.
1. We should try to keep the 1 file - 1 sample ratio for ease of including those snippets in various places.

Turns out, achieving the 1:1 ratio would require a lot of code being replicated in multiple places with only small adjustments. As a result, keeping it all up to date would soon become a big chore. Changing one thing may require changing many different files.

#### So how does this PR help?
Since I noticed that most of the samples share a lot of common code, I decided to write a script that will assemble the code samples from pieces - this way, the code duplication in all the samples will be under control and updating things should be much simpler. Now, with the SGS we have a clear set of ingredients and recipes that can be used to create new code snippets with minimal need for new code.

## Note: The code samples will be transformed into the new format in following PR.

### CODE REVIEW

File: samples/sgs.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+    output_file = []
+    header_added = False
+    for line in file_lines:
+        if IMPORTS_FILL.search(line):
+            output_file.append(import_lines)
+        elif match := INGREDIENT_FILL.search(line):
+            output_file.append(ingredients[match.group(1)].text)
+        elif REGION_START.search(line):
+            output_file.append(REGION_START.sub("# [START \\1]", line))
```
Comment (FrodoTheTrue) [inline samples/sgs.py:224]: We should not to forget to ignore region tag here. Are there any ways how to disable it (and in README also)?

File: samples/sgs.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        for file in sorted(unknown_files):
+            print(f" - {repr(file)}")
+
+
+def verify(args: argparse.Namespace):
+    # TODO: Needs to check if the files are up to date. Will be used to auto-check every commit.
+    pass
+
+
+def parse_arguments():
```
Comment (FrodoTheTrue) [inline samples/sgs.py:309]: Let's add to the README file input params doc and example of usage.
Example: https://github.com/GoogleCloudPlatform/compute-migration-scripts/tree/main/instances-to-stateful-mig#usage

File: samples/README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+To add a new ingredient, create a new `.py` file with the code you want to later use in the snippets. Mark the beginning
+of the code you want to include with `# <INGREDIENT name>` and the end with `# </INGREDIENT>`. 
+
+Please leave the imports required by this ingredient **OUTSIDE** the area marked with ingredient comments. The SGS 
+script will automatically collect all the required imports and put them in the final snippet in the right place and in
+right order.
+
+## Preparing a recipe
+Each recipe is a file located in the `recipes` folder. It should have the `.py` extension and should be a valid Python 
+script. Each recipe has to have an `# <IMPORTS/>` line and at least one `# <INGREDIENT name/>` line.
```
Comment (FrodoTheTrue) [inline samples/README.md:43]: Why at least one INGREDIENT is obligated? What if I want to store all samples in recipes, but some of them are independent?

Probably we should allow users to add some imports in recipes and them merge all imports: from recipes and from ingredients.

**Comment by FrodoTheTrue** [comment]
Overall looks really great! It's a good step to reusing code in the samples (really strong problem for now).

Some thoughts:
1. We definitely should connect this script with other languages (and make specific repo for the script). I see the problem only with imports (but I believe it could be solved by using extensions for current parser).
2. Another reason to spread this script throughout is readability. I believe code will become more "functional" (but maybe I am wrong)
Before:
````
[START 1]
a = ...
[END 1]
```
After:
[START 1]
def do():
    ....
a = do()
[END 1]
``` 
Before variants look more simple and preferable from users readability perspective.
Except of functions I see now one good use case: not to write wait operation code every time.

3. WDYT about renaming recipes/ingredients to templates/blocks? Probably could be a bit simpler.

File: samples/sgs.py (hunk 1 above)
Comment (m-strzelczyk) [reply samples/sgs.py:224]: I honestly don't know... When I wrote this, I thought that region tag has to be at the beginning of the line. Now the snippet bot is complaining... I guess I'll need to do something silly and break it up so it doesn't match some regex.

File: samples/sgs.py (hunk 2 above)
Comment (m-strzelczyk) [reply samples/sgs.py:309]: That's really pretty :P I'll make something similar I guess.

**Comment by m-strzelczyk** [comment]
> 1. We definitely should connect this script with other languages (and make specific repo for the script). I see the problem only with imports (but I believe it could be solved by using extensions for current parser).

Yes, that's the plan. I think once this turns out to be useful, we would make a tool that can have language-specific "plugins" to parse and compose files from different languages.



> 2\. Another reason to spread this script throughout is readability. I believe a final snippet code will become more "functional" (but maybe I am wrong) and should decide which way we prefer.

I'm afraid I don't understand what you're saying here. Can you perhaps give a bigger example?



> 3\. WDYT about renaming recipes/ingredients to templates/blocks? Probably could be a bit simpler.

I think we could switch the naming to `templates` and `blocks` or `parts`. I don't really mind the names.

File: samples/sgs.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        file_content = file.read()
+    # Read imports
+    simple_imports, imports_from = parse_imports(file_content)
+    # Read the script
+    for line in file_content.splitlines(keepends=True):
+        if in_ingredient and INGREDIENTS_END.match(line):
+            break
+        elif in_ingredient:
+            ingredient_lines.append(line)
+        elif match := INGREDIENTS_START.match(line):
```
Comment (parthea) [inline samples/sgs.py:132]:
```
nox > flake8 --show-source --builtin=gettext --max-complexity=20 --import-order-style=google '--exclude=.nox,.cache,env,lib,generated_pb2,*_pb2.py,*_pb2_grpc.py' --ignore=E121,E123,E126,E203,E226,E24,E266,E501,E704,W503,W504,I202 --max-line-length=88 --application-import-names test_sgs,sgs_test_fixtures,snippets,noxfile,sgs,.nox .
[./sgs.py:132](https://cs.corp.google.com/piper///depot/google3/sgs.py?l=132):21: E999 SyntaxError: invalid syntax
        elif match := INGREDIENTS_START.match(line):
                    ^
```

**Review by parthea — APPROVED** [review]

**Comment by parthea** [comment]
The samples tests are failing with
```
    from flaky import flaky
E   ModuleNotFoundError: No module named 'flaky'
```

**Pull request merged**